File: src/readers/gtfs/schedule/calendar.ts
import { parseCSVAsRecord } from '../../';
import { parseGtfsDate } from '../utils';

/**
 * Enumeration to represent day availability in the calendar.
 * 0 = Not available, 1 = Available
 */
export enum GTFSDayAvailability {
  /** 0 - Service not available on this day. */
  NotAvailable = 0,
  /** 1 - Service available on this day. */
  Available = 1,
}

/**
 * # Calendar Information
 *
 * **Conditionally Required**
 * Defines a set of dates when service is available for one or more routes.
 * Required unless all dates of service are defined in `calendar_dates.txt`.
 */
export class GTFSCalendar {
  /**
   * **Required**
   * Identifies a set of dates when service is available.
   */
  serviceId: string;
  /**
   * **Required**
   * Service availability on Mondays: 0 or 1.
   */
  monday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Tuesdays: 0 or 1.
   */
  tuesday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Wednesdays: 0 or 1.
   */
  wednesday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Thursdays: 0 or 1.
   */
  thursday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Fridays: 0 or 1.
   */
  friday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Saturdays: 0 or 1.
   */
  saturday: GTFSDayAvailability;
  /**
   * **Required**
   * Service availability on Sundays: 0 or 1.
   */
  sunday: GTFSDayAvailability;
  /**
   * **Required**
   * Start service day (inclusive) for the interval. Format: YYYYMMDD
   */
  startDate: Date;
  /**
   * **Required**
   * End service day (inclusive) for the interval. Format: YYYYMMDD
   */
  endDate: Date;

  /** @param data - the parsed GTFS CSV data */
  constructor(data: Record<string, string>) {
    this.serviceId = data.service_id;
    this.monday =
      data.monday !== undefined
        ? (parseInt(data.monday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.tuesday =
      data.tuesday !== undefined
        ? (parseInt(data.tuesday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.wednesday =
      data.wednesday !== undefined
        ? (parseInt(data.wednesday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.thursday =
      data.thursday !== undefined
        ? (parseInt(data.thursday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.friday =
      data.friday !== undefined
        ? (parseInt(data.friday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.saturday =
      data.saturday !== undefined
        ? (parseInt(data.saturday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.sunday =
      data.sunday !== undefined
        ? (parseInt(data.sunday, 10) as GTFSDayAvailability)
        : GTFSDayAvailability.NotAvailable;
    this.startDate = parseGtfsDate(data.start_date);
    this.endDate = parseGtfsDate(data.end_date);
  }
}

/**
 * @param input - the input string to parse from
 * @returns - an array of Calendars
 */
export function parseGTFSCalendars(input: string): GTFSCalendar[] {
  const data = parseCSVAsRecord(input);
  return data.map((d) => new GTFSCalendar(d));
}

File: rust/readers/gtfs/schedule/calendar.rs
use crate::{
    readers::{csv::parse_csv_as_record, parse_gtfs_date},
    util::Date,
};
use alloc::{string::String, vec::Vec};
use s2json::MValueCompatible;

/// Enumeration to represent day availability in the calendar.
/// 0 = Not available, 1 = Available
#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
pub enum GTFSDayAvailability {
    /// 0 - Service not available on this day.
    NotAvailable = 0,
    /// 1 - Service available on this day.
    Available = 1,
}
impl From<&str> for GTFSDayAvailability {
    fn from(s: &str) -> Self {
        match s.trim() {
            "1" => GTFSDayAvailability::Available,
            _ => GTFSDayAvailability::NotAvailable,
        }
    }
}

/// # Calendar Information
///
/// **Conditionally Required**
/// Defines a set of dates when service is available for one or more routes.
/// Required unless all dates of service are defined in `calendar_dates.txt`.
#[derive(Debug, Default, Clone, PartialEq, MValueCompatible)]
pub struct GTFSCalendar {
    /// **Required**
    /// Identifies a set of dates when service is available.
    pub service_id: String,
    /// **Required**
    /// Service availability on Mondays: 0 or 1.
    pub monday: String,
    /// **Required**
    /// Service availability on Tuesdays: 0 or 1.
    pub tuesday: String,
    /// **Required**
    /// Service availability on Wednesdays: 0 or 1.
    pub wednesday: String,
    /// **Required**
    /// Service availability on Thursdays: 0 or 1.
    pub thursday: String,
    /// **Required**
    /// Service availability on Fridays: 0 or 1.
    pub friday: String,
    /// **Required**
    /// Service availability on Saturdays: 0 or 1.
    pub saturday: String,
    /// **Required**
    /// Service availability on Sundays: 0 or 1.
    pub sunday: String,
    /// **Required**
    /// Start service day (inclusive) for the interval. Format: YYYYMMDD
    pub start_date: String,
    /// **Required**
    /// End service day (inclusive) for the interval. Format: YYYYMMDD
    pub end_date: String,
}
impl GTFSCalendar {
    /// Create a new GTFSCalendar
    pub fn new(source: &str) -> Vec<GTFSCalendar> {
        let mut res = Vec::new();
        for record in parse_csv_as_record::<GTFSCalendar>(source, None, None) {
            res.push(record);
        }
        res
    }
    /// Get the availability for Monday
    pub fn monday(&self) -> GTFSDayAvailability {
        self.monday.as_str().into()
    }
    /// Get the availability for Tuesday
    pub fn tuesday(&self) -> GTFSDayAvailability {
        self.tuesday.as_str().into()
    }
    /// Get the availability for Wednesday
    pub fn wednesday(&self) -> GTFSDayAvailability {
        self.wednesday.as_str().into()
    }
    /// Get the availability for Thursday
    pub fn thursday(&self) -> GTFSDayAvailability {
        self.thursday.as_str().into()
    }
    /// Get the availability for Friday
    pub fn friday(&self) -> GTFSDayAvailability {
        self.friday.as_str().into()
    }
    /// Get the availability for Saturday
    pub fn saturday(&self) -> GTFSDayAvailability {
        self.saturday.as_str().into()
    }
    /// Get the availability for Sunday
    pub fn sunday(&self) -> GTFSDayAvailability {
        self.sunday.as_str().into()
    }
    /// Get the start date
    pub fn start_date(&self) -> Date {
        parse_gtfs_date(&self.start_date).unwrap_or_default()
    }
    /// Get the end date
    pub fn end_date(&self) -> Date {
        parse_gtfs_date(&self.end_date).unwrap_or_default()
    }
}
